src/taskgraph/optimize/base.py

```diff
@@ -307,7 +328,9 @@ def replace_tasks(
             deadline = max(
                 resolve_timestamps(now, task.task["deadline"]) for task in dependents
             )
-        repl = opt.should_replace_task(task, params, deadline, arg)
+        repl = opt.should_replace_task(
+            task, params, deadline, arg, label_to_taskid, taskid_to_status
+        )
```

Comment:
This will break custom optimization strategies defined in project specific code.

That in and of itself isn't a huge deal (we can do a major version bump if necessary).. but these values are specific to the `IndexSearch` strategy, so I don't love that we would pass them in to every strategy.

I'm not sure what the best alternative here is however..

There is a way to pass arguments into the strategies' constructor, but that happens in the decorator at module import time, so we wouldn't have `target_task_graph` yet.

The other option could be to pass the necessary data in via classmethod / classproperty. E.g, we could set `target_task_graph` as a class property on all strategies, then `IndexSearch` can use that to lazily make the calls to Taskcluster (and memoize the result). That feels a bit cleaner to me.

Reply:
I agree that this is suboptimal. I don't really understand your last paragraph; why would we use a `classmethod/classproperty` specifically and why would that help to pass down the arguments?

I've pushed a hacky version of passing conditionally the arguments so we can clear the rest of the pending comments.